proxy/vm_test.go

```
@@ -0,0 +1,157 @@
+	// the end of Stop().
+	rig.stopFds, err = rig.detector.Snapshot()
+	assert.Nil(rig.t, err)
+
+	assert.True(rig.t,
+		rig.detector.Compare(os.Stdout, rig.startFds, rig.stopFds))
+}
+
+const testVM = "testVM"
+
+// CreateVM creates a vm instance that is connected to the rig's Hyperstart
+// mock object.
+func (rig *vmRig) CreateVM() *vm {
+	ctlSocketPath, ioSocketPath := rig.Hyperstart.GetSocketPaths()
+
+	vm := newVM(testVM, ctlSocketPath, ioSocketPath)
+	assert.NotNil(rig.t, vm)
+
+	err := vm.Connect()
+	assert.Nil(rig.t, err)
+
+	return vm
+}
+
+// CreateIOConn returns a connection and an ioBase sequence number that can be
+// used to send and receive IO data to/from hyperstart through the vm object.
+// This function will allocate n consecutive sequence numbers.
+func (rig *vmRig) CreateIOConn(vm *vm, n int) (net.Conn, uint64) {
+	// this mocks a shim connected to the proxy, we'll give the proxy end
+	// of the socketpair() to allocateIO while returning the shim end the
+	// for the test code  to use.
+	shimConn, proxyConn, err := Socketpair()
+	assert.Nil(rig.t, err)
+
+	seq := vm.AllocateIo(n, 1, proxyConn)
+
+	return shimConn, seq
+}
+
+// TestShimGone tests what happens when a shim goes away and we still have I/O
+// data we'd like to write to it. This happens when the shim is forcefully
+// killed by SIGKILL, for instance (docker kill)
+func TestShimGone(t *testing.T) {
+	rig := NewVMRig(t)
+	rig.Start()
+
+	vm := rig.CreateVM()
+	shim, seq := rig.CreateIOConn(vm, 1)
+
+	// Simulate the shim going away.
+	shim.Close()
+
+	// Simulate the VM process sending data for a shim that is no longer there.
+	rig.Hyperstart.SendIoString(seq, "foo\n")
+
+	// Wait until the vm goroutine has read the IO data from hyperstart
+	// sent in the line above.
+	// Because that goroutine should cleanup vm.ioSessions on error, we
+	// use that condition for the wait. There isn't really an easier to
+	// synchronise here.
```

I'm still confused - this is test code, so you can assert the behaviour, so why not? The comment says "`Because that goroutine should cleanup vm.ioSessions on error...`" but what if that behaviour doesn't happen?

All I'm suggesting is that you add a check to ensure that `len(vm.ioSessions) == 0` when that loop has exited and error if not since it appears that that block could fail, but it would be a silent failure.